interop-tests/src/arch.rs

```diff
@@ -177,35 +209,40 @@ pub(crate) mod wasm {
         futures_timer::Delay::new(duration).boxed()
     }
 
-    pub(crate) fn build_transport(
-        local_key: Keypair,
+    pub(crate) async fn build_swarm<B: NetworkBehaviour>(
         ip: &str,
         transport: Transport,
-    ) -> Result<(BoxedTransport, String)> {
+        behaviour_constructor: impl FnOnce(&Keypair) -> B,
+    ) -> Result<(Swarm<B>, String)> {
         match transport {
-            Transport::Webtransport => Ok((
-                libp2p::webtransport_websys::Transport::new(
-                    libp2p::webtransport_websys::Config::new(&local_key),
-                )
-                .boxed(),
-                format!("/ip4/{ip}/udp/0/quic/webtransport"),
-            )),
-            Transport::WebRtcDirect => Ok((
-                webrtc::Transport::new(webrtc::Config::new(&local_key)).boxed(),
-                format!("/ip4/{ip}/udp/0/webrtc-direct"),
-            )),
+            Transport::Webtransport => {
+                let swarm = libp2p::SwarmBuilder::with_new_identity()
+                    .with_wasm_bindgen()
+                    .with_other_transport(|key| {
+                        libp2p::webtransport_websys::Transport::new(
+                            libp2p::webtransport_websys::Config::new(key),
+                        )
+                        .map(|(peer_id, conn), _| (peer_id, StreamMuxerBox::new(conn)))
```

Comment:
Should we do this `.map` for the user inside the builder?

Reply:
I am hesitant to do so. What if the transport already returns a `StreamMuxerBox`?

Reply:
> I am hesitant to do so. What if the transport already returns a `StreamMuxerBox`?

Are you worried about an additional allocation? Or the indirection caused by another pointer look-up?

I think this would be a nice usability improvement. It is a problem that users are facing, for example: https://github.com/libp2p/rust-libp2p/issues/3179

This is only present for obscure transport combinations which usually only happens with alpha-releases of our transports. Unless there are concerns other than the above performance implications, I'd say that the usability is more important.

Reply:
It would be great if we could tag this PR as resolving https://github.com/libp2p/rust-libp2p/issues/3179 :)